### PR TITLE
Establish connection with fhir gwan

### PR DESCRIPTION
**Pull request by thanakritTW**
mostly create and interface for  repository to get patients

details in the commits

PR against record_discovery branch for the purpose of code review because we rebased this branch

### CODE REVIEW

File: src/In.ProjectEKA.HipService/OpenMrs/OpenMrsClient.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+using System.Threading.Tasks;
+
+namespace In.ProjectEKA.HipService.OpenMrs
+{
+    public class OpenMrsClient: IOpenMrsClient
+    {
+        private readonly HttpClient httpClient;
+        private readonly OpenMrsConfiguration configuration;
+
+        public OpenMrsClient(HttpClient httpClient, OpenMrsConfiguration openMrsConfiguration)
```
Comment (davidviktorTW) [inline src/In.ProjectEKA.HipService/OpenMrs/OpenMrsClient.cs:15]: ...and the dependencies like HttpClient

File: test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+            // Disable SSL verification in test only
+            var handler = new HttpClientHandler()
+            {
+                ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
+            };
+            var httpClient = new HttpClient(handler);
+            var openmrsConfiguration = new OpenMrsConfiguration {
+                Url = "https://bahmni-0.92.bahmni-covid19.in/openmrs/",
+                Username = "superman",
+                Password = "Admin123"
```
Comment (davidviktorTW) [inline test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs:117]: It's public repo :)

File: src/In.ProjectEKA.HipService/wwwroot/openapi.yaml [hunk 3]
```diff
@@ -18,7 +18,7 @@ tags:
   - name: consents
 
 paths:
-  /patients/discover/carecontexts:
+  /patients/discover/carecontexts/test:
```
[the remark this answers is not in the document]
Comment (thanakritTW) [reply src/In.ProjectEKA.HipService/wwwroot/openapi.yaml:21]: oh noo, will fix this

File: src/In.ProjectEKA.HipService/OpenMrs/OpenMrsClient.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace In.ProjectEKA.HipService.OpenMrs
+{
+    public class OpenMrsClient: IOpenMrsClient
```
[the remark this answers is not in the document]
Comment (thanakritTW) [reply src/In.ProjectEKA.HipService/OpenMrs/OpenMrsClient.cs:10]: yes, will do after making sure it is okay to put default username and password on appsettings.json

File: test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs (hunk 2 above)
Comment (thanakritTW) [reply test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs:117]: :( It has default username and password on public [wiki](https://bahmni.atlassian.net/wiki/spaces/BAH/pages/14712841/Bahmni+Virtual+Box) as well , please remind us to make sure this is squashed after fixing and the branch is deleted

Comment (thanakritTW) [reply test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs:117]: also the server is protected by firewall

File: src/In.ProjectEKA.HipService/OpenMrs/DiscoveryDataSource.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+        public DiscoveryDataSource(IOpenMrsClient openMrsClient)
+        {
+            this.openMrsClient = openMrsClient;
+        }
+
+        public async Task<List<Patient>> LoadPatientsAsync(string name, AdministrativeGender? gender, string yearOfBirth)
+        {
+            var path = DiscoveryPathConstants.OnPatientPath;
+            var query = HttpUtility.ParseQueryString(string.Empty);
```
Comment (praveenadayanand) [inline src/In.ProjectEKA.HipService/OpenMrs/DiscoveryDataSource.cs:23]: Is this how typically an empty collection is initialised?

**Comment by thanakritTW** [comment]
> c



> @thanakritTW Is it correct that you want to merge this branch with record_discovery branch?

No, the moment I created this branch  record_discovery is not merged yet. So it is better to close this and use https://github.com/Bahmni-Covid19/hip-service/pull/4

File: src/In.ProjectEKA.HipService/OpenMrs/DiscoveryDataSource.cs (hunk 5 above)
Comment (thanakritTW) [reply src/In.ProjectEKA.HipService/OpenMrs/DiscoveryDataSource.cs:23]: I would say empty httpQueryString object

File: src/In.ProjectEKA.HipService/OpenMrs/DiscoveryPathConstants.cs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+﻿namespace In.ProjectEKA.HipService.OpenMrs
+{
+    public static class DiscoveryPathConstants
+    {
+        public const string OnPatientPath = "ws/fhir2/Patient";
```
[the remark this answers is not in the document]
Comment (thanakritTW) [reply src/In.ProjectEKA.HipService/OpenMrs/DiscoveryPathConstants.cs:5]: @morgan-tw

**Comment by praveenadayanand** [comment]
Closing this PR as a new PR was created for the same card.

File: src/In.ProjectEKA.HipService/OpenMrs/DiscoveryPathConstants.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (thanakritTW) [reply src/In.ProjectEKA.HipService/OpenMrs/DiscoveryPathConstants.cs:5]: In the GatewayPathConstants.cs, many constants has 'On' prefixed, we would like to make it consistent.

File: src/In.ProjectEKA.HipService/OpenMrs/DiscoveryDataSource.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+                query["gender"]=gender.ToString().ToLower();
+            }
+            if (!string.IsNullOrEmpty(yearOfBirth)) {
+                query["birthdate"]=yearOfBirth;
+            }
+            if (query.ToString() != ""){
+                path = $"{path}/?{query}";
+            }
+
+            var patients = new List<Patient>();
```
[the remark this answers is not in the document]
Comment (morgan-tw) [reply src/In.ProjectEKA.HipService/OpenMrs/DiscoveryDataSource.cs:37]: @praveenadayanand I understood that the story only concern the happy path 

We have no clue what to do if the connection fail. 
Of course if we can log it but also it is business cases like should we notify the patient etc ...
This also why we suggested the HealthCheck.

So what should we do exactly, just logging the status code and the according message ?

**Pull request closed without merging**